epytope/TCRSpecificityPrediction/External.py

```diff
@@ -709,4 +709,118 @@ def format_results(self, filenames, tcrs, epitopes, pairwise):
+            "VDJ_cdr3": "TCRB_CDR3"
+        }
+        required_columns = list(rename_columns.values()) + ["Epitope"]
+        df_tcrs = tcrs.to_pandas(rename_columns=rename_columns)
+        if pairwise:
+            df_tcrs = self.combine_tcrs_epitopes_pairwise(df_tcrs, epitopes)
+        else:
+            df_tcrs = self.combine_tcrs_epitopes_list(df_tcrs, epitopes)
+        model_type = "B" if "model_type" not in kwargs else kwargs["model_type"]
+        if model_type == "B":
+            df_tcrs = self.filter_by_length(df_tcrs, None, "TCRB_CDR3", "Epitope")
+            df_tcrs = df_tcrs[(~df_tcrs["TCRB_CDR3"].isna()) & (df_tcrs["TCRB_CDR3"] != "")]
+            prediction_columns = ["TCRB_CDR3", "Epitope"]
+        elif model_type == "A":
+            df_tcrs = self.filter_by_length(df_tcrs, "TCRA_CDR3", None, "Epitope")
+            df_tcrs = df_tcrs[(~df_tcrs["TCRA_CDR3"].isna()) & (df_tcrs["TCRA_CDR3"] != "")]
+            prediction_columns = ["TCRA_CDR3", "Epitope"]
+        elif model_type == "AB":
+            df_tcrs = self.filter_by_length(df_tcrs, "TCRA_CDR3", "TCRB_CDR3", "Epitope")
+            df_tcrs = df_tcrs[(~df_tcrs["TCRB_CDR3"].isna()) & (df_tcrs["TCRB_CDR3"] != "")]
+            df_tcrs = df_tcrs[(~df_tcrs["TCRA_CDR3"].isna()) & (df_tcrs["TCRA_CDR3"] != "")]
+            prediction_columns = ["TCRB_CDR3", "TCRA_CDR3", "Epitope"]
+        else:
+            raise TypeError(f"Please specify a correct model_type: A, B or AB")
+        df_tcrs = df_tcrs[required_columns]
+        df_tcrs.drop_duplicates(subset=prediction_columns, inplace=True, keep="first")
+        df_tcrs = df_tcrs.reset_index(drop=True)
+        return df_tcrs
+    
+    def save_tmp_files(self, data, **kwargs):
+        tmp_folder = self.get_tmp_folder_path()
+        model_type = "B" if "model_type" not in kwargs else kwargs["model_type"]
+        path_in = os.path.join(tmp_folder.name, f"{self.name}_input.xlsx")
+        path_out = os.path.join(tmp_folder.name, f"TCR{model_type}_pred.csv")
+        data.to_excel(path_in)
+        return [path_in, path_out], tmp_folder
+
+    def get_base_cmd(self, filenames, tmp_folder, interpreter=None, conda=None, cmd_prefix=None, **kwargs):
+        model_type = "B" if "model_type" not in kwargs else kwargs["model_type"]
+        cmd_epitope = ["from Model_Predict_Feature_Extraction import *",
+                       "from DLpTCR_server import *",
+                       f"error_info,TCRA_cdr3,TCRB_cdr3,Epitope = deal_file('''{filenames[0]}''','''{tmp_folder.name}/''','''{model_type}''')",
+                       f"output_file_path = save_outputfile('''{tmp_folder.name}/''', '''{model_type}''', '''{filenames[0]}''',TCRA_cdr3,TCRB_cdr3,Epitope)"
+                       ]
+        cmd_epitope = f'python -c "{"; ".join(cmd_epitope)}"'
+        return cmd_epitope
+    
+    def run_exec_cmd(self, cmd, filenames, interpreter=None, conda=None, cmd_prefix=None, **kwargs):
+        cmds = []
```

Comment:
please adapt to the new run_exec_cmd. It runs better across platform. (eg. the conda init error from the cluster does not appear with that).

Reply:
done, but, please, check it.